stock_by_warehouse/models/product_product.py

```diff
@@ -0,0 +1,195 @@
+import json
+from collections import defaultdict
+
+from odoo import _, api, fields, models
+from odoo.tools import float_is_zero
+
+UNIT = "Product Unit of Measure"
+
+
+class ProductProduct(models.Model):
+    _inherit = "product.product"
+
+    qty_available_not_res = fields.Float(
+        string="Qty Available Not Reserved",
+        digits=UNIT,
+        compute="_compute_qty_available_not_reserved",
+    )
+
+    warehouses_stock = fields.Text(
+        store=False,
+        readonly=True,
+    )
+
+    warehouses_stock_location = fields.Text(store=False, readonly=True)
+
+    warehouses_stock_recompute = fields.Boolean(store=False)
+
+    @api.onchange("warehouses_stock_recompute")
+    def _warehouses_stock_recompute_onchange(self):
+        if not self.warehouses_stock_recompute:
+            self.warehouses_stock_recompute = True
+            return
+        self.warehouses_stock = self._compute_get_quantity_warehouses_json()
+        self.warehouses_stock_location = self._compute_get_stock_location()
+        self.warehouses_stock_recompute = True
+
+    def _product_available_not_res_hook(self, quants):
+        """Hook used to introduce possible variations"""
+        return False
+
+    def _prepare_domain_available_not_reserved(self):
+        domain_quant = [("product_id", "in", self.ids)]
+        domain_quant_locations = self._get_domain_locations()[0]
+        domain_quant.extend(domain_quant_locations)
+        return domain_quant
+
+    @api.depends_context("warehouse", "company")
+    def _compute_qty_available_not_reserved(self):
+        """Method to set the quantity available that is not reserved of a serie of products."""
+        domain_quant = self._prepare_domain_available_not_reserved()
+        quants = (
+            self.env["stock.quant"]
+            .with_context(lang=False)
+            ._read_group(
+                domain_quant,
+                groupby=["product_id"],
+                aggregates=["quantity:sum", "reserved_quantity:sum"],
+            )
+        )
+
+        res = {}
+        for quant in quants:
+            quantity = quant[1] - quant[2]
+            res[quant[0].id] = max(quantity, 0.0)
```

Comment:
These are harder to read.

Maybe:

`for product, quantity, reserved_quantity in quants:`

In fact, now using dict comprehension could be clearer

Reply:
done

Reply:
By dict comprehension, I meant:

```
res = {
    product.id: max(quantity - reserved_quantity, 0.0)
    for product, quantity, reserved_quantity in quants
}
```

That would be faster because creating a filled dictionary is faster than creating an empty one and updating one key a time.

Reply:
done, sorry i miss understood.